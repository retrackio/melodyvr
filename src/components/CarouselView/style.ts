import styled from 'styled-components';

export const Container = styled.div`
  max-width: 100%;
  margin: 0 auto;
  overflow: hidden;
`;

export const Row = styled.div`
  width: 0;
  height: 100%;
  transition: transform 0.5s ease-in-out;
`;

export const Item = styled.div`
  float: left;
  box-sizing: border-box;
  padding: 10px;
`;
