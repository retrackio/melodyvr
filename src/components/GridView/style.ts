import styled from 'styled-components';

export const Container = styled.div`
  max-width: 100%;
  margin: 0 auto;
`;

export const Row = styled.div`
  display: flex;
  flex-flow: row wrap;
  justify-content: flex-start;
`;

export const Item = styled.div`
  flex-basis: 25%;
  box-sizing: border-box;
  /* position: relative; */
  padding: 10px;

  @media (max-width: 1024px) {
    flex-basis: 33.3%;
  }

  @media (max-width: 768px) {
    flex-basis: 50%;
  }

  @media (max-width: 600px) {
    flex-basis: 100%;
  }
`;
