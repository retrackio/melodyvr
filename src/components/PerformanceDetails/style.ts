import styled from 'styled-components';

export const Title = styled.p`
  font-family: 'Proza Libre';
  font-weight: 600;
  font-size: 1.8rem;
  color: #000;
  margin-bottom: 20px;
  margin-top: 50px;
`;

export const Description = styled.p`
  font-family: 'Open Sans';
  font-weight: 400;
  font-size: 1.6rem;
  color: #000;
  margin-bottom: 20px;
`;

export const Label = styled.p`
  font-family: 'Open Sans';
  font-weight: 400;
  font-size: 1.4rem;
  color: #000;
  margin-bottom: 20px;
`;

export const Debug = styled.pre`
  font-family: 'Consolas,monospace';
  font-size: 1.2em;
  overflow-x: auto;
  height: 250px;
  border: 1px solid #999;
  padding: 20px;
  margin: 20px 0px;
`;
