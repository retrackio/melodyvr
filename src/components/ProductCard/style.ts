import styled from 'styled-components';

interface WrapperProps {
  accentColor: string;
  image: string;
}

interface TitleProps {
  backgroundColor: string;
}

export const Container = styled.div`
  position: relative;
`;

export const Link = styled.a`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  z-index: 2;
`;

export const Wrapper = styled.div<WrapperProps>`
  box-sizing: initial;
  height: 250px;
  width: 100%;
  overflow: hidden;
  position: relative;
  border-top: 6px solid ${(props) => `#${props.accentColor}`};
  background-image: ${(props) => `url(${props.image})`};
  background-size: cover;
  background-position: 50% 50%;
`;

export const Title = styled.h2<TitleProps>`
  font-family: 'Proza Libre';
  font-weight: 600;
  font-size: 1.8rem;
  color: #fff;
  position: absolute;
  left: 0px;
  top: 0px;
  padding: 10px;
  background-color: ${(props) => `#${props.backgroundColor}`};
  box-shadow: 2px 2px 12px rgba(0, 0, 0, 0.4);
`;

export const Description = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 10px;
  line-height: 1.8rem;
  color: #fff;
  background: rgba(0, 0, 0, 0.5);
  font-size: 1.3rem;
`;
