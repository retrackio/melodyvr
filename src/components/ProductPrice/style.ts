import styled from 'styled-components';

export const Container = styled.p`
  position: absolute;
  right: 0;
  top: 0;
  font-size: 1.4rem;
  padding: 5px 10px;
  background: rgba(255, 255, 255, 0.9);
`;
