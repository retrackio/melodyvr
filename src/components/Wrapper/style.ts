import styled, { createGlobalStyle } from 'styled-components';

export const GlobalStyle = createGlobalStyle`
  html {
    font-size: 62.5%;
  }
`;

export const Header = styled.header`
  padding: 20px;
  background: #000;
  color: #fff;
  font-family: 'Proza Libre';
  font-weight: 600;
  font-size: 1.4rem;
  display: flex;
  align-items: center;
`;

export const HeaderTitle = styled.h1`
  font-size: 2.2rem;
  flex-grow: 1;
  margin-right: 10px;
  & svg {
    height: 30px;
  }
`;

export const MainContainer = styled.main`
  max-width: 1280px;
  margin: 0 auto;
  padding: 5px;
  background: #fff;
  color: #000;
  font-family: 'Open Sans';
  font-weight: 400;
`;
