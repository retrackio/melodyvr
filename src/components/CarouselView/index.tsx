import React, { useState, useRef, useEffect } from 'react';
import debounce from 'lodash.debounce';
import * as Styled from './style';

interface Props {
  items: React.ReactNode[];
}

export const RESIZE_DEBOUNCE_MS = 50;

const CarouselView = ({ items }: Props) => {
  const [currentItem, setCurrentItem] = useState<number>(0);
  const [availableWidth, setAvailableWidth] = useState<number>(0);
  const containerRef = useRef<HTMLElement>(null);

  const totalItems = items.length;
  const hasNextSlide = currentItem + 1 < totalItems;
  const hasPrevSlide = currentItem > 0;
  const rowWidth = availableWidth * totalItems;
  const rowOffset = -1 * availableWidth * currentItem;

  useEffect(() => {
    const handleResize = debounce(() => {
      setAvailableWidth(containerRef.current.offsetWidth);
    }, RESIZE_DEBOUNCE_MS);
    setAvailableWidth(containerRef.current.offsetWidth);

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return (
    <>
      <Styled.Container ref={(ref) => (containerRef.current = ref)}>
        <Styled.Row
          style={{
            width: rowWidth,
            transform: `translateX(${rowOffset}px)`,
          }}
        >
          {items.map((item, index) => (
            <Styled.Item
              style={{ width: availableWidth }}
              key={`carousel_item_${index}`}
            >
              {item}
            </Styled.Item>
          ))}
        </Styled.Row>
      </Styled.Container>
      <button
        name="prev"
        disabled={!hasPrevSlide}
        onClick={() => setCurrentItem(currentItem - 1)}
      >
        Prev
      </button>
      <button
        name="next"
        disabled={!hasNextSlide}
        onClick={() => setCurrentItem(currentItem + 1)}
      >
        Next
      </button>
    </>
  );
};

export default CarouselView;
