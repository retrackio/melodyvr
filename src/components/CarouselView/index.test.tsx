import React from 'react';
import { shallow, mount } from 'enzyme';
import debounce from 'lodash.debounce';
import { act } from 'react-dom/test-utils';
import CarouselView, { RESIZE_DEBOUNCE_MS } from './';
import * as Styled from './style';

const items = [<div key="1">item 1</div>, <div key="2">item 2</div>];

jest.mock('lodash.debounce', () =>
  jest.fn().mockImplementation((fnc: Function) => () => fnc()),
);

describe('<CarouselView/>', () => {
  beforeEach(() => {
    Object.defineProperties(window.HTMLElement.prototype, {
      offsetWidth: {
        get() {
          return 10;
        },
      },
    });
  });

  it('renders correctly', () => {
    const wrapper = shallow(<CarouselView items={items} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calculates and applies the correct width for the container and each item inside it', () => {
    const wrapper = mount(<CarouselView items={items} />);
    expect(wrapper.find(Styled.Row).prop('style').width).toEqual(20);
    wrapper
      .find(Styled.Item)
      .forEach((item) => expect(item.prop('style').width).toEqual(10));
  });

  describe('Resizing', () => {
    it('applies a new width once the window resizes', () => {
      const wrapper = mount(<CarouselView items={items} />);
      expect(wrapper.find(Styled.Row).prop('style').width).toEqual(20);
      Object.defineProperties(window.HTMLElement.prototype, {
        offsetWidth: {
          get() {
            return 50;
          },
        },
      });
      act(() => {
        ((global as unknown) as Window).dispatchEvent(new Event('resize'));
      });
      wrapper
        .update()
        .find(Styled.Item)
        .forEach((item) => expect(item.prop('style').width).toEqual(50));
      expect(wrapper.find(Styled.Row).prop('style').width).toEqual(100);
      expect(debounce).toHaveBeenCalledTimes(2);
    });

    it('adds the resize listener once mounted', () => {
      const originalAddEventListener = window.addEventListener;
      window.addEventListener = jest.fn();
      mount(<CarouselView items={items} />);
      expect(window.addEventListener).toHaveBeenCalledWith(
        'resize',
        expect.any(Function),
      );
      expect(debounce).toHaveBeenCalledWith(
        expect.any(Function),
        RESIZE_DEBOUNCE_MS,
      );
      window.addEventListener = originalAddEventListener;
    });

    it('removes the resize listener once the component unmounts', () => {
      const originalRemoveEventListener = window.removeEventListener;
      window.removeEventListener = jest.fn();
      const wrapper = mount(<CarouselView items={items} />);
      wrapper.unmount();
      expect(window.removeEventListener).toHaveBeenCalledWith(
        'resize',
        expect.any(Function),
      );
      window.removeEventListener = originalRemoveEventListener;
    });
  });

  describe('Buttons', () => {
    it('moves the items correctly when "next" and "prev" are clicked', () => {
      const wrapper = mount(<CarouselView items={items} />);
      wrapper.find('button[name="next"]').simulate('click');
      expect(wrapper.find(Styled.Row).prop('style').transform).toEqual(
        'translateX(-10px)',
      );
      wrapper.find('button[name="prev"]').simulate('click');
      expect(wrapper.find(Styled.Row).prop('style').transform).toEqual(
        'translateX(0px)',
      );
    });

    it('disables the buttons correctly', () => {
      const wrapper = mount(<CarouselView items={items} />);
      const expectButtonDisabledStateToBe = (name: string, state: boolean) => {
        expect(wrapper.find(`button[name="${name}"]`).prop('disabled')).toEqual(
          state,
        );
      };
      expectButtonDisabledStateToBe('next', false);
      expectButtonDisabledStateToBe('prev', true);
      wrapper.find('button[name="next"]').simulate('click');
      expectButtonDisabledStateToBe('next', true);
      expectButtonDisabledStateToBe('prev', false);
    });
  });
});
