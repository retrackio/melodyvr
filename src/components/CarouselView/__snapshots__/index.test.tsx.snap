// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarouselView/> renders correctly 1`] = `
<Fragment>
  <style__Container>
    <style__Row
      style={
        Object {
          "transform": "translateX(0px)",
          "width": 0,
        }
      }
    >
      <style__Item
        key="carousel_item_0"
        style={
          Object {
            "width": 0,
          }
        }
      >
        <div
          key="1"
        >
          item 1
        </div>
      </style__Item>
      <style__Item
        key="carousel_item_1"
        style={
          Object {
            "width": 0,
          }
        }
      >
        <div
          key="2"
        >
          item 2
        </div>
      </style__Item>
    </style__Row>
  </style__Container>
  <button
    disabled={true}
    name="prev"
    onClick={[Function]}
  >
    Prev
  </button>
  <button
    disabled={false}
    name="next"
    onClick={[Function]}
  >
    Next
  </button>
</Fragment>
`;
